Add KingTriumphant, AquaRider, OverloadCluster (dm06)

DanieloV opened the pull request:
+ Add the cards KingTriumphant, AquaRider, OverloadCluster from dm06. Added them all together since they have the exact same effect. (Used this for rulings https://duelmasters.fandom.com/wiki/Aqua_Rider/Rulings)
+ Add an event that triggers after a card is played and all its effects applied. This is handled a bit different then the other events in that I'm passing the match player that played the card as well. Looks a bit ugly, so I'm up for suggestions if you think there is a better way to get which player played the card. I've seen this handled in other parts by searching different containers for the card but that seems quite cumbersome and prone to unexpected future effects.

sindreslungaard commented:
In general we should try to avoid adding new events unless absolutely necessary as the way the event system is implemented is (regrettably) not that performant if we keep on adding new events because all events are being handled by every card even if the event is not relevant for the card (a better way would have been if a card can subscribe to only certain events)

But that aside, can this not be solved by simply listening for the CardPlayedEvent and if it's the opponents card that is played, give our card blocker?

DanieloV commented:
> In general we should try to avoid adding new events unless absolutely necessary as the way the event system is implemented is (regrettably) not that performant if we keep on adding new events because all events are being handled by every card even if the event is not relevant for the card (a better way would have been if a card can subscribe to only certain events)
> 
> But that aside, can this not be solved by simply listening for the CardPlayedEvent and if it's the opponents card that is played, give our card blocker?

It can not, at least in a simple way. I tried initially to use CardPlayedEvent and SpellCast (as spells don't use CardPlayedEvent). The problem is there are cards that trigger after those are called. 

Like for example Scarlet Skyterror that activates on CardMoved. This triggers the card to get blocker and then get affected by the effect from Scarlet, when it should get the blocker only after Scarlet's effect. 
Also it wasn't working with Shield Triggers properly like Critical Blade, as the order was again reversed.

That's why I ended up initially adding an event to let me know when a card was played and all it's effects activated, as it handles all those unexpected situations, though I understand the issue with adding more events as well. 
There could maybe be a solution that checks for CardMoved, which might be less clean, but I can look into it if you prefer that to adding another event.

sindreslungaard commented:
You're right, it would have to be applied on CardMoved and SpellCast events instead of CardPlayed. I think that might be necessary anyways so that all effects that trigger from summons/card play are from the same event(s) so that if there's a card in the future that has something like "ignore effects that trigger when this card is summoned" it would be able to cancel this one out - we might even have some cards that cancels out effects from CardMoved already that would not work wiith a new event like this.

At some point I think we should refactor all cards to use CardPlayed instead of CardMoved and SpellCast like they do now and also implement an optional priority number to the "ScheduleAfter" function so that it becomes easier to deal with the order of execution from card to card.

For now though it's probably best to keep adding cards like this in the same way others have been done (CardMoved and SpellCast) to avoid issues in the future and to also make it easier to refactor to a better order of execution system in the future

DanieloV commented:
Ok, this is ready for another check.
 
I modified this to use CardMoved and SpellCast. It works as expected including with Scarlet Skyterror and with Shield triggers.

I removed the new event, but I modified the 2 ones above to send the match player id, I think this way it's less error prone then trying to figure out where a card ended up and looking for it in that place.
